### PR TITLE
Change XSUB -> XPUB message processing.

### PR DESCRIPTION
**Pull request by drolevar**
Now if the first frame in a multipart message is not subscribe/unsubscribe,
the rest of the parts are also considered to be not subscribe/unsubscribe.

### CODE REVIEW

**Comment by drolevar** [comment]
@somdoron @mrseanmorris @bluca Ok, can this be a solution to your concerns?
Should I still look into making it a socket option?

**Comment by bluca** [comment]
If you want to add a new socket option to change the behaviour, that's fine. Also please fix the formatting:

```
--- a/src/xsub.cpp
+++ b/src/xsub.cpp
@@ -104,13 +104,11 @@ int zmq::xsub_t::xsend (msg_t *msg_)
     bool first_part = !_more_send;
     _more_send = (msg_->flags () & msg_t::more) != 0;
 
-    if (first_part)
-    {
+    if (first_part) {
         _first_subscribe = false;
-    }
-    else if (!_first_subscribe)
+    } else if (!_first_subscribe)
         //  User message sent upstream to XPUB socket
-        return _dist.send_to_all(msg_);
+        return _dist.send_to_all (msg_);
 
     if (msg_->is_subscribe () || (size > 0 && *data == 1)) {
         //  Process subscribe message
```

**Comment by drolevar** [comment]
@bluca If the change will be accepted into the master and stay there for subsequent releases, it's fine for me to keep it as it is.
But I'm not the project maintainer. Is it you making strategic decisions for the project?

**Comment by sigiesec** [comment]
@drolevar One small thing: Could you please also update the commit message to follow the "Problem"/"Solution" convention as described in http://wiki.zeromq.org/docs:contributing section "Good commit messages"? Thanks!

The documentation in https://github.com/zeromq/libzmq/blob/master/doc/zmq_socket.txt should be updated to clarify this behaviour.

Regarding the socket option: Strictly, this is a breaking change, since the existing documentation could be interpreted in a way that showed the previous behaviour. However, IMO it seems rather unlikely that someone relied on this, so maybe it's fine without a socket option. In case someone runs into this and deems it a regression, a socket option might be added later on.

**Comment by somdoron** [comment]
I agree with @sigiesec, it is a breaking change, but if it only after a none-sub/unsub I think we are fine without a socket option. @bluca what do you think?

**Comment by drolevar** [comment]
@sigiesec Fixed.
@bluca What do you think?

**Comment by bluca** [comment]
Sorry, I'll have to be the bad cop: there are a gazillion users doing the strangest things. In other words, if there's a chance it might break something, it will - please let's add a socket option

**Comment by drolevar** [comment]
@bluca It's OK to be a bad cop. Someone has to :)

**Comment by drolevar** [comment]
@bluca @sigiesec @somdoron
OK, socket option is implemented. Please, review and comment.

File: include/zmq.h [hunk 1]
```diff
@@ -381,6 +381,7 @@ ZMQ_EXPORT const char *zmq_msg_gets (const zmq_msg_t *msg_,
 #define ZMQ_GSSAPI_PRINCIPAL_NAMETYPE 90
 #define ZMQ_GSSAPI_SERVICE_PRINCIPAL_NAMETYPE 91
 #define ZMQ_BINDTODEVICE 92
+#define ZMQ_ONLY_FIRST_SUBSCRIBE 93
```
Comment (bluca) [inline include/zmq.h]: 93 already exists - needs to be added in the ZMQ_BUILD_DRAFT_API below, and duplicated in src/zmq_drafts.hpp

File: tests/test_xpub_manual.cpp [hunk 2]
```diff
@@ -458,50 +458,57 @@ void test_user_message ()
 
 void test_user_message_multi ()
 {
+    const int only_first_subscribe = 1;
+
     //  Create a publisher
     void *pub = test_context_socket (ZMQ_XPUB);
     TEST_ASSERT_SUCCESS_ERRNO (zmq_bind (pub, "inproc://soname"));
+    TEST_ASSERT_SUCCESS_ERRNO (zmq_setsockopt (pub, ZMQ_ONLY_FIRST_SUBSCRIBE,
```
Comment (bluca) [inline tests/test_xpub_manual.cpp]: the test body needs to be wrapper in ```#ifdef ZMQ_ONLY_FIRST_SUBSCRIBE``` since it will be DRAFT for a while

Comment (drolevar) [reply tests/test_xpub_manual.cpp]: @bluca I guess it should the same in the xsub.cpp and xpub.cpp?

File: include/zmq.h (hunk 1 above)
Comment (drolevar) [reply include/zmq.h]: Should be fixed now.

File: tests/test_xpub_manual.cpp (hunk 2 above)
Comment (drolevar) [reply tests/test_xpub_manual.cpp]: Should be fixed now.

File: include/zmq.h (hunk 1 above)
Comment (bluca) [reply include/zmq.h]: It's still missing from src/zmq_drafts.hpp though? With that, there's no need to ifdef in the implementation, but only for external usage

**Pull request merged**